[receiver/mongodb] Add mongodb host resource attributes

dmitryax opened the pull request:
The new attribute are added to the MongoDB receiver to distinguish metrics coming from different MongoDB instances.
  - `server.address`: The address of the MongoDB host, enabled by default.
  - `server.port`: The port of the MongoDB host, disabled by default.

Resolves https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32350 and https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32810

crobert-1 commented on .chloggen/mongodb_add_address_res_attr.yaml:
```diff
@@ -0,0 +1,30 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: enhancement
```
Out of ignorance: It's considered a breaking change when metrics are switched to be enabled by default. Since we're adding a resource attribute that is now enabled by default, should this also be a breaking change? Or is it so little extra data that it doesn't merit the breaking label?

I guess counterpoint, adding attributes to metrics isn't considered a breaking change, so I'm not sure where the line in the sand is. (I'm also not finding documentation on this, either way).

dmitryax replied:
We don't have strict guidelines for the resource attributes. I'd say adding a new resource attribute is not a breaking change because it doesn't change cardinality. Removing or changing one should be considered a breaking change, though.

Happy to hear other opinions. cc @djaglowski

djaglowski replied:
I've generally considered it a breaking change because it breaks time series that are based on all attributes.

dmitryax replied:
By breaking time series you mean just the interrupted continuity, no other user-facing effect, right?

dmitryax replied:
I would merge this as is, assuming that's the case ^.  Let me know if there are any other consequences for the end user that I miss, I'll change it.

crobert-1 commented on .chloggen/mongodb_add_address_res_attr.yaml:
```diff
@@ -0,0 +1,30 @@
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: |
+  The new attribute are added to the MongoDB receiver to distinguish metrics coming from different MongoDB instances.
```
```suggestion
  The new resource attributes are added to the MongoDB receiver to distinguish metrics coming from different MongoDB instances.
```
Nit: pluralize `attributes`, and also clarify that they're `resource` attributes.

crobert-1 commented on receiver/mongodbreceiver/scraper.go:
```diff
@@ -22,6 +24,8 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver/internal/metadata"
 )
 
+const defaultMongodbPort = 27017
```
There's an existing `mongoPort` variable in `integration_test.go` that we can remove as a result of this addition. (Since the `integration_test.go` variable is hidden behind the `integration` build flag, it would be better to use this one.)

Also, could we use this variable [when creating the default config](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/93a78c6ad8f57c5e4ff29a5f41779af03fc85145/receiver/mongodbreceiver/factory.go#L34)?

dmitryax replied:
That's a string

crobert-1 replied:
Right, but we could just create a temporary string variable and use it instead. Not a big deal, just trying to deduplicate wherever possible.

```
defaultEndpoint := fmt.Sprintf("%s:%v", "localhost", defaultMongodbPort)
```

dmitryax replied:
In both cases, we need port, not endpoint: int64 and string. I wouldn't personally bother with the conversion here. I don't think it simplifies things or reduces the code.

dmitryax replied:
Oh, I was talking about the the integration tests only where we define `mongoPort`. The `defaultEndpoint` probably makes sense. Let me apply that

crobert-1 replied:
That's fair, I didn't realize `mongoPort` was a string. This is fine with me.

crobert-1 replied:
> Oh, I was talking about the integration tests only where we define mongoPort. The defaultEndpoint probably makes sense. Let me apply that

Yeah, sorry, I misinterpreted your initial response to be about the second part of my comment, and my `defaultEndpoint` comment was responding to that part. 🙂

crobert-1 commented on receiver/mongodbreceiver/scraper_test.go:
```diff
@@ -382,3 +382,61 @@ func TestTopMetricsAggregation(t *testing.T) {
+				"host": "localhost",
+			},
+			expectedAddress: "localhost",
+			expectedPort:    27017,
```
```suggestion
			expectedPort:    defaultMongodbPort,
```
Nit: Use central variable

crobert-1 commented on receiver/mongodbreceiver/scraper.go:
```diff
@@ -202,3 +213,23 @@ func (s *mongodbScraper) recordAdminStats(now pcommon.Timestamp, document bson.M
 	s.recordIndexAccess(now, indexStats, databaseName, collectionName, errs)
 }
+
+func serverAddressAndPort(serverStatus bson.M) (string, int64, error) {
```
Would it be helpful to have named return arguments here? Would it be more clear if we named the method `getServerAddressAndPort`? This is pretty simple method, so neither of these things are that big of a deal, just slightly more clear on purpose and functionality.

_Just some nits, feel free to drop if they're not helpful._

dmitryax replied:
> Would it be helpful to have named return arguments here?

I think the name of the function server**Address**And**Port** plays the role of named return arguments here

> Would it be more clear if we named the method getServerAddressAndPort? 

get prefix is an anti-pattern in go :) https://go.dev/doc/effective_go#Getters

crobert-1 replied:
Fair point on the return arguments clarity. Thanks for sharing the docs reference too, I'm definitely wrong there 😅

crobert-1 reviewed (commented):
Just some nits, the general logic looks good to me though.

dmitryax committed:
[receiver/mongodb] Add mongodb host resource attributes

The new attribute are added to the MongoDB receiver to distinguish metrics coming from different MongoDB instances.
  - `server.address`: The address of the MongoDB host, enabled by default.
  - `server.port`: The port of the MongoDB host, disabled by default.

Co-authored-by: Curtis Robert <crobert@splunk.com>

crobert-1 commented:
_Looks like the integration expected test data needs to be updated to include the new resource attribute._
